node/src/components/linear_chain/state.rs

```diff
@@ -191,8 +142,16 @@ impl<I> LinearChain<I> {
             .remove(&public_key, &block_hash)
     }
 
-    // Caches the signature.
-    pub(super) fn cache_signatures(&mut self, signatures: BlockSignatures) {
+    /// Caches the signature.
+    pub(super) fn cache_signatures(&mut self, mut signatures: BlockSignatures) {
+        // Merge already-known signatures and the new ones.
+        self.get_signatures(&signatures.block_hash)
+            .iter()
+            .for_each(|bs| {
```

Comment:
This would be less confusing:
```suggestion
        if let Some(bs) = self.get_signatures(&signatures.block_hash) {
```